sycl/source/detail/graph_impl.cpp

```diff
@@ -1002,138 +1017,309 @@ exec_graph_impl::~exec_graph_impl() {
+
+  if (EventNeeded) {
+    return SchedulerEvent;
+  }
+  return nullptr;
+}
+
+EventImplPtr exec_graph_impl::enqueuePartitionWithScheduler(
+    std::shared_ptr<partition> &Partition,
+    const std::shared_ptr<sycl::detail::queue_impl> &Queue,
+    sycl::detail::CG::StorageInitHelper CGData, bool EventNeeded) {
+
+  if (!Partition->MRequirements.empty()) {
+    CGData.MRequirements.insert(CGData.MRequirements.end(),
+                                Partition->MRequirements.begin(),
+                                Partition->MRequirements.end());
+    CGData.MAccStorage.insert(CGData.MAccStorage.end(),
+                              Partition->MAccessors.begin(),
+                              Partition->MAccessors.end());
+  }
+
+  auto CommandBuffer = Partition->MCommandBuffers[Queue->get_device()];
+
+  std::unique_ptr<sycl::detail::CG> CommandGroup =
+      std::make_unique<sycl::detail::CGExecCommandBuffer>(
+          CommandBuffer, nullptr, std::move(CGData));
+
+  EventImplPtr SchedulerEvent = sycl::detail::Scheduler::getInstance().addCG(
+      std::move(CommandGroup), Queue, EventNeeded);
+
+  if (EventNeeded) {
+    SchedulerEvent->setEventFromSubmittedExecCommandBuffer(true);
+    return SchedulerEvent;
+  }
+
+  return nullptr;
+}
+
+EventImplPtr exec_graph_impl::enqueuePartitionDirectly(
+    std::shared_ptr<partition> &Partition,
+    const std::shared_ptr<sycl::detail::queue_impl> &Queue,
+    std::vector<detail::EventImplPtr> &WaitEvents, bool EventNeeded) {
+
+  auto CheckURResult = [](ur_result_t UrResult) {
+    if (UrResult != UR_RESULT_SUCCESS) {
+      throw sycl::exception(
+          errc::event, "Failed to enqueue event for command buffer submission");
+    }
+  };
+
+  // Create a list containing all the UR event handles in WaitEvents. WaitEvents
+  // is assumed to be safe for scheduler bypass and any host-task events that it
+  // contains can be ignored.
+  std::vector<ur_event_handle_t> UrEventHandles{};
+  UrEventHandles.reserve(WaitEvents.size());
+  for (auto &SyclWaitEvent : WaitEvents) {
+    auto URHandle = SyclWaitEvent->getHandle();
+    if (URHandle) {
+      UrEventHandles.push_back(URHandle);
+    }
+  }
+
+  auto CommandBuffer = Partition->MCommandBuffers[Queue->get_device()];
+  const size_t UrEnqueueWaitListSize = UrEventHandles.size();
+  const ur_event_handle_t *UrEnqueueWaitList =
+      UrEnqueueWaitListSize == 0 ? nullptr : UrEventHandles.data();
+
+  if (!EventNeeded) {
+    ur_result_t UrResult =
+        Queue->getAdapter()
+            ->call_nocheck<sycl::detail::UrApiKind::urEnqueueCommandBufferExp>(
+                Queue->getHandleRef(), CommandBuffer, UrEnqueueWaitListSize,
+                UrEnqueueWaitList, nullptr);
+    CheckURResult(UrResult);
+    return nullptr;
+  } else {
     auto NewEvent = std::make_shared<sycl::detail::event_impl>(Queue);
     NewEvent->setContextImpl(Queue->getContextImplPtr());
     NewEvent->setStateIncomplete();
+    NewEvent->setSubmissionTime();
```

Comment:
Can we create the `NewEvent` after the call to `urEnqueueCommandBufferExp`? Or would it conflict with setting the correct submission time?